### ISSUE TITLE
Shap.plots.bar and shap.plots.waterfall - Are these working currently?

### ISSUE THREAD

**Issue by chetanambi** [shap/shap#1606]
Hi - 

### Issue# 1
I am following the example plot for for bar and waterfall [here](https://shap.readthedocs.io/en/latest/example_notebooks/plots/bar.html) but not able to run the code. Can anyone confirm are these deprecated in the latest versions of SHAP.

```
import xgboost
import shap

# train XGBoost model
X,y = shap.datasets.adult()
model = xgboost.XGBClassifier().fit(X, y)

# compute SHAP values
explainer = shap.Explainer(model, X)
shap_values = explainer(X)
```

---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
/usr/local/lib/python3.6/dist-packages/shap/explainers/_linear.py in supports_model_with_masker(model, masker)
    268         try:
--> 269             Linear._parse_model(model)
    270         except e:

3 frames
/usr/local/lib/python3.6/dist-packages/shap/explainers/_linear.py in _parse_model(model)
    255         else:
--> 256             raise Exception("An unknown model type was passed: " + str(type(model)))
    257 

Exception: An unknown model type was passed: <class 'xgboost.sklearn.XGBClassifier'>

During handling of the above exception, another exception occurred:

NameError                                 Traceback (most recent call last)
<ipython-input-115-ae84e18a2a15> in <module>()
      7 
      8 # compute SHAP values
----> 9 explainer = shap.Explainer(model, X)
     10 shap_values = explainer(X)

/usr/local/lib/python3.6/dist-packages/shap/explainers/_explainer.py in __init__(self, model, masker, link, algorithm, output_names, feature_names, **kwargs)
     97 
     98                 # use implementation-aware methods if possible
---> 99                 if explainers.Linear.supports_model_with_masker(model, self.masker):
    100                     algorithm = "linear"
    101                 elif explainers.Tree.supports_model_with_masker(model, self.masker): # TODO: check for Partition?

/usr/local/lib/python3.6/dist-packages/shap/explainers/_linear.py in supports_model_with_masker(model, masker)
    268         try:
    269             Linear._parse_model(model)
--> 270         except e:
    271             return False
    272         return True

NameError: name 'e' is not defined

### **Issue 2:**
When I try to use plots.bar and plots.waterfall with TreeExplainer, I get below error:
`shap.plots.bar(shap_values_train)`

---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-117-6c4261a1920b> in <module>()
----> 1 shap.plots.bar(shap_values_train)

/usr/local/lib/python3.6/dist-packages/shap/plots/_bar.py in bar(shap_values, max_display, order, clustering, clustering_cutoff, merge_cohorts, show_data, show)
     49         cohorts = shap_values.cohorts
     50     else:
---> 51         assert isinstance(shap_values, dict), "You must pass an Explanation object, Cohorts object, or dictionary to bar plot!"
     52 
     53     # unpack our list of Explanation objects we need to plot

AssertionError: You must pass an Explanation object, Cohorts object, or dictionary to bar plot!

**Note: I am using Colab**

**Comment by mrektor** [comment]
Hi, I am not able to run the code either...

Python 3.8.3
Shap version: 0.37.0

I get the same error

**Comment by XiangBu** [comment]
hi, have u solved the problem, I met the same problem when running this  https://github.com/slundberg/shap/blob/master/notebooks/general/Explainable%20AI%20with%20Shapley%20Values.ipynb

**Comment by hiyamgh** [comment]
I am getting:
```
shap.plots.waterfall(shap_values[1]) 
```
TypeError                                 Traceback (most recent call last)
<ipython-input-15-8da6b4b86207> in <module>
----> 1 shap.plots.waterfall(shap_values[1])

TypeError: 'module' object is not callable

**Comment by Z-Shakeri** [comment]
I had the same problem, replaced the shap.values with explainer(X) and it worked.

**Comment by VickyVouk** [comment]
> I had the same problem, replaced the shap.values with explainer(X) and it worked.

Could you please provide the code detailed? I can't fix this either. Thank you very very much in advance
@slundberg

**Comment by phinguyen44** [comment]
I'm having the same issue. Upvoting for visibility.

**Comment by ozziejin** [comment]
same issue here,

TypeError: 'module' object is not callable

**Comment by wundermahn** [comment]
Has this issue ever been resolved?

**Comment by Z-Shakeri** [comment]
> > I had the same problem, replaced the shap.values with explainer(X) and it worked.
> 
> Could you please provide the code detailed? I can't fix this either. Thank you very very much in advance
> @slundberg

In this line: shap.plots.bar(shap_values_train), I replaced the `shap_values_train` parameter with `explainer(X)`. I hope this solves your problem too. I will re-run my script and will update this answer if anything changes.

**Comment by Mark531** [comment]
It seems that waterfall has become a module, and the function to be called is waterfall_plot. But curiously, it's not specified that way in the docs.

**Comment by alexandrerays** [comment]
For me

This:
```
ex = shap.TreeExplainer(model)

ex(X)
```
resulted in:

> ValueError: could not convert string to float: 'MY_CATEGORY_NAME'

---

and this:
```
ex = shap.TreeExplainer(model, X)
```
resulted in

> Found a NULL input array in _cext_dense_tree_update_weights!
> 
> ...
> 
> ValueError: could not convert string to float: 'MY_CATEGORY_NAME'
> ...
> 
> Exception: Currently TreeExplainer can only handle models with categorical splits when feature_perturbation="tree_path_dependent" and no background data is passed. Please try again using shap.TreeExplainer(model, feature_perturbation="tree_path_dependent").

---

Same errors using `tree_path_dependent` and `Explainer` instead of `TreeExplainer`.

Anyone resolved?

**Comment by medomatto** [comment]
I have the same issue. Anyone?

**Comment by JuanMLE** [comment]
I have the same issue too.

**Comment by OriKatz1** [comment]
I also have similar issue

**Comment by loicmagne** [comment]
Same issue this library looks broken . . .

**Comment by LukeValley** [comment]
@Z-Shakeri 
> I had the same problem, replaced the shap.values with explainer(X) and it worked.

Please provide a code sample, still not working for me.

**Comment by BaiqingL** [comment]
Given some dataframe with column `text` that contains a sentence.
```python
cv = CountVectorizer(max_features=4000)
X = cv.fit_transform(sentimentTrain.text)
y = sentimentTrain.sentiment
model = xgboost.XGBClassifier().fit(X, y)
explainer = shap.Explainer(model, X)
```
Causes that issue.

**Comment by gigios10** [comment]
I have the same issue too. Anyone solved?

**Comment by talentingsun** [comment]
shap.plots.bar(explainer(x_test))

**Comment by tomaszek0** [comment]
The problem solution:
shap.summary_plot(shap_values[1], X_test, plot_type="bar")

**Comment by alexmorphine** [comment]
I had the same issue with shap 0.42.1, so instead of passing shap_values I created a new object (you can skip clustering and names):
`explanation = shap.Explanation(shap_values, clustering=clustering, feature_names=out_cols)`
and only then
`shap.plots.bar(explanation, clustering=clustering)`

**Comment by joelostblom** [comment]
The two issues described in the OP both work for me on shap 0.43.0, so maybe this has been resolved and can be closed?

**Comment by connortann** [comment]
Closing as this issue appears to be fixed since we merged in the various fixes from the "Community fork" in #2943.

Please don't hesitate to open a new issue if you encounter this again.
